transcoder/src/storage/storage.go

```diff
@@ -0,0 +1,74 @@
+package storage
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+)
+
+type ContentsWriterCallback func(ctx context.Context, writer io.Writer) error
+
+// Handles storage and retrieval of items in a storage backend.
+// "Items" are pieces of data that can be stored and retrieved.
+// Paths with ".." are not allowed, and may result in unexpected behavior.
+type StorageBackend interface {
+	// DoesItemExist checks if an item exists in the storage backend.
+	DoesItemExist(ctx context.Context, path string) (bool, error)
+	// ListItemsWithPrefix returns a list of items in the storage backend that match the given prefix.
+	// Note: returned items may have a "/" in them, e.g. "foo/bar/baz".
+	ListItemsWithPrefix(ctx context.Context, pathPrefix string) ([]string, error)
+	// DeleteItem deletes an item from the storage backend. If the item does not exist, it returns nil.
+	// Deletion should be "syncronous" (i.e. the function should block until the write is complete).
+	DeleteItem(ctx context.Context, path string) error
+	// SaveItemWithCallback saves an item to the storage backend. If the item already exists, it overwrites it.
+	// The writeContents function is called with a writer to write the contents of the item.
+	// Writes should be "syncronous" (i.e. the function should block until the write is complete).
+	SaveItemWithCallback(ctx context.Context, path string, writeContents ContentsWriterCallback) error
```

Comment:
I think using a callback here makes it harder to reason about the code. Wouldn't this work:

```go
Save(ctx context.Context, path string, content io.Reader) error
```

`io.Reader` is the parameter of the parameter that takes aws & would work with an `io.Copy` for local fs.

Reply:
I can do this, but the tradeoff is that anything that expects an io.writer (such as `imaging.Encode`) will need to re-implement reader/writer piping as well as concurrency (because the reader and writer need to be read/written to at the same time).

What would you think about two functions: one that takes a writer callback, and one that takes a simple io.Reader? Something like `SaveItemWithWriter` and SaveItemWithReader`, or `SaveItemWithCallback` and `SaveItem`?

Reply:
That's a good point, didn't think about `imaging.Encode`. Having both `Save` & `SaveWithCallback` sounds good to me!